### ISSUE TITLE
JDL App generation is broken

### ISSUE THREAD

**Issue by deepu105** [jhipster/generator-jhipster#7504]
The new JDL app generation support is broken, below are the issues

- [ ] The parser expects, appType, DB and basename to be passed, which is not logical for app generation as it needs to be inferred from the app config
- [ ] The import-jdl sub gen expects a .yo-rc.json when generating applications, it should be skipped
- [ ] The `path` option is not required and might cause issues(especially in JH online), it needs to be removed as we really don't need it, and for microservices, we could always create folders from the current directory
- [ ] When the app config is created it needs to call the app generator to create the actual source code as well

cc @MathieuAA

**Comment by MathieuAA** [comment]
Didn't the merged PR cover these?Le 19 avr. 2018 14:05, Deepu K Sasidharan <notifications@github.com> a écrit :The new JDL app generation support is broken, below are the issues
 The parser expects, appType, DB and basename to be passed, which is not logical for app generation as it needs to be inferred from the app config The import-jdl sub gen expects a .yo-rc.json when generating applications, it should be skipped The path option is not required and might cause issues(especially in JH online), it needs to be removed as we really don't need it, and for microservices, we could always create folders from the current directory When the app config is created it needs to call the app generator to create the actual source code as well
cc @MathieuAA

**Comment by deepu105** [comment]
@MathieuAA I don't think so

**Comment by deepu105** [comment]
I was trying this
```
application {
  config {
    baseName MyApp
    applicationType monolith
    databaseType sql
    prodDatabaseType mysql
  }
}
```

**Comment by MathieuAA** [comment]
I won't be able to properly look into it until the 24th. But this smells weird.Le 19 avr. 2018 17:42, Deepu K Sasidharan <notifications@github.com> a écrit :I was trying this
application {
  config {
    baseName MyApp
    applicationType monolith
    databaseType sql
    prodDatabaseType mysql
  }
}

**Comment by jdubois** [comment]
I confirm we still have the issue - as JHipster 5 is just around the corner, shouldn't we better remove the option to generate apps for the moment? @MathieuAA do you have time to fix this, or can we help you?

**Comment by deepu105** [comment]
I would still prefer to fix this as the issues are easily fixable and its a
great feature.

I would also like to remove the path option in the JDL as its not necessary
for our workflows and it would cause lot of issues for us (by users who
will abuse it)

I can also do this once I'm back from vacation if Matheue is busy

**Comment by MathieuAA** [comment]
@jdubois I'm fixing these issues as I write this. @deepu105 I'll remove it after fixing the app generation.
What does "around the corner" mean? :)

**Comment by jdubois** [comment]
@MathieuAA we have a "JHipster Day" at Ippon on Tuesday with @PierreBesson @pascalgrimaud @wmarques and my trainees so it should be a good day to release! We'll need to have some Opencollective drinks an evening with you and @cbornet (but not on Tuesday it's the 10th anniversary of the ParisJUG).

**Comment by deepu105** [comment]
@jdubois i would suggest to push back the release a bit as well. There are stuff that needs to be ironed out, like e2e tests for React etc.
Also, there are some issues to be fixed with the Blueprint system which is a blocker for v5 IMO.
I'm on vacation this month so it will be difficult for me to work on that stuff.

I guess a perfect day for the release would be the JHipster conf, we could even do it during the Keynote :)

**Comment by jdubois** [comment]
I understand but we've been delaying the release for about one month now, and everybody is waiting for our React support. For me we need to release ASAP, and then we'll improve.

- For React e2e: I know they don't work, but those are not a blocker, we can release without them
- Blueprints: yes that's very annoying, but nobody uses them at the moment, so they can be postponed
- JDL: I don't really know the status here. Is that easy to rollback? Is there any issue with staying with JDL v1? JDL v1 was working very well, there's no need to wait for JDL v2 to do a release

-> for me we should do a v5 next week, and then all of these will for v5.1, v5.2, v5.3... If we need to wait for everything we'll never release

**Comment by MathieuAA** [comment]
Quite easy. I've just finished implementing a fix for the application generation, I'm currently writing the tests. Don't hold your breath for me, there are quite a lot of tests to move around or write.

**Comment by deepu105** [comment]
Technically we dont have to roll back as there is no breaking changes in
JDL. We just need to remove docs may be untill its fixed

Thanks & Regards,
Deepu

**Comment by MathieuAA** [comment]
@deepu105 @jdubois I'll release JCore v3 today or tomorrow (and create a PR for the subgen)

**Comment by jdubois** [comment]
Awesome, thanks @MathieuAA !! I'll do a new beta release just afterwards

**Comment by deepu105** [comment]
​​
awesome, you can assign me as a reviewer once PR is done

Thanks & Regards,
Deepu

**Comment by MathieuAA** [comment]
Okay, the last PR didn't work. I'm fixing it like Deepu suggested and dropping a PR.

**Comment by MathieuAA** [comment]
@pascalgrimaud with the `with-entitiies` flag, I don't have to compose three generators anymore, just jdl & entity or jdl & app. Thanks :)

**Comment by pascalgrimaud** [comment]
nice @MathieuAA !!

**Comment by MathieuAA** [comment]
@pascalgrimaud you told me a few days ago that you managed to reproduce the issue by using the `--with entities` flag, and no yarn link. What's your setup to test without linking?

**Comment by MathieuAA** [comment]
@pascalgrimaud nevermind!!! I found the issue. Yeoman **has** to declare options before anything else (it won't pick up the config file otherwise). I don't know why.
Look, this doesn't work:
```js
module.exports = class extends BaseGenerator {
    constructor(args, opts) {
        super(args, opts);

        this.configOptions = ths.options.configOptions || {};
        // This adds support for a `--skip-client` flag
        this.option('skip-client', {
            desc: 'Skip the client-side application generation',
            type: Boolean,
            defaults: false
        });
```
And if I don't initialize `this.configOptions` up top, but at the bottom, it works.
Must be a special property of JS class constructors, or a side-effect of declaring for the first time options for Yeoman. This sucks.
At least, I'm fixing this today.

**Comment by pascalgrimaud** [comment]
In fact, I managed to reproduce the issue without doing anything:
- copy a sample `.yo-rc.json` with some entities inside `.jhipster/` (BankAccount, Label, Operation)
- type `jhipster --with-entities`
But it's resolved now.

Currently, I can't do a `yarn link generator-jhipster` as I'm with Yarn 1.7.0 and there is an issue with it. But I do it after the 1st generator instead of before.

About Travis:
- Travis uses this for installing JHipster: https://github.com/jhipster/generator-jhipster/blob/master/travis/scripts/00-install-jhipster.sh#L72
- there is no `yarn link` as it doesn't work well with Travis
- you can do the same locally if you unlink your generator-jhipster

**Comment by MathieuAA** [comment]
@pascalgrimaud That's what I did, I did what the script does. It succeeds beautifully or fails miserably each time I do a `yarn global add file:...`, in turns.

**Comment by MathieuAA** [comment]
@pascalgrimaud Nuking the cache, removing links and trying again seems to be consistent, and works. Thanks ;)

**Comment by deepu105** [comment]
Btw you can also do npm link to do the linking and the use yarn for
everything else

Thanks & Regards,
Deepu

**Comment by MathieuAA** [comment]
@deepu105 Yes :)
